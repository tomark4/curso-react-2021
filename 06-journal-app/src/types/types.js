export const types = {
  // auth  types
  login: '[Auth] LOGIN',
  logout: '[Auth] LOGOUT',
  // ui types
  uiAddError:'[ui] add error',
  uiRemoveError:'[ui] delete error',
  uiStartLoading: '[ui] start loading',
  uiStopLoading: '[ui] stop loading',
}
