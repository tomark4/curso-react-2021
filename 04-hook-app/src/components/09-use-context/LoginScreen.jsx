import { useContext } from "react"
import {UserContext} from "./UserContextProvider"

const LoginScreen = () => {

    const {setUser} = useContext(UserContext);

    return (
        <>
            <h1>Login</h1>
            <button className="btn btn-primary"
            onClick={() => setUser({id: 1, name:'jose',email:'jotta@jotta.com'})}>Ingresar</button>

            <button className="btn btn-danger"
            onClick={() => setUser({})}>
                Logout
            </button>
        </>
    )
}

export default LoginScreen
