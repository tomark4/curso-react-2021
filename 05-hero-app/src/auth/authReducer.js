import { types } from '../types/types'

const authReducer = (state = {}, action) => {
  // reducers para manejar el login
  switch(action.type){
    case types.login:
      return {...action.payload, logged: true}

    case types.logout:
      return { logged: false }

    default:
      return state
  }
}

export default authReducer;
