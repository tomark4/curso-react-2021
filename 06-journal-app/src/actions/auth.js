import { types } from '../types/types';
import {firebase, googleAuthProvider} from '../firebase';
import { startLoading, stopLoading } from './ui';
import Swal from 'sweetalert2';
import { noteLogout } from './notes';

// login with email and password
export const loginWithEmailPassword = (email, password) => (dispatch) => {
  dispatch( startLoading() );
  firebase.auth().signInWithEmailAndPassword(email, password)
  .then( ({user}) =>  {
    dispatch({
      type: types.login,
      payload: { uid: user.uid, name: user.displayName }
    });
    dispatch( stopLoading() );
  }).catch(err => {
    console.log(err);
    Swal.fire('Error',err.message, 'error')
    dispatch( stopLoading() );
  });
}

// login con google
export const loginWithGoogle = () => (dispatch) => {
  firebase.auth().signInWithPopup( googleAuthProvider )
  .then( ({user}) => {
    dispatch({
      type: types.login,
      payload: {uid: user.uid, name: user.displayName}
    })
  })
}

// register con email and password
export const registerWithEmailPassword = (email, password, name) => (dispatch) => {
  firebase.auth().createUserWithEmailAndPassword(email, password)
  .then( async ({user}) => {
    await user.updateProfile({displayName: name});
    dispatch({ type: types.login, payload: {uid: user.uid, name: user.displayName} });
  }).catch( err => console.log(err) );
}


export const startLogout = () => async (dispatch) => {
  try{
    await firebase.auth().signOut();
    dispatch( noteLogout() );
    dispatch( logout() );
  }catch(err){
    console.log(err)
  }
}

export const logout = () => ({
  type: types.logout
})
