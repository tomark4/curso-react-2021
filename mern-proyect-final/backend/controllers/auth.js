const { response } = require('express');
const Usuario = require("../models/Usuario");
const bcrypt = require("bcryptjs");
const { generateJWT } = require("../helpers/jwt");

const register = async (req, res = response) => {
  try{
    const { name, email, password } = req.body;

    // validar que el email no exista
    let usuario = await Usuario.findOne({email});

    if(usuario){
      return res.status(400).json({
        ok: false,
        message: 'usuario existe con ese correo'
      });
    }

    // crear si no existe
    usuario = new Usuario({name, email, password});

    // encriptar la constraseña
    const salt = bcrypt.genSaltSync();
    usuario.password = bcrypt.hashSync( password, salt);
    await usuario.save();

    const token = await generateJWT(usuario.id, usuario.name);

    return res.status(201).json({
      ok: true,
      uid: usuario.id,
      name: usuario.name,
      message: 'user created',
      token
    });

  } catch(err){
    console.log(err);
    return res.status(500).json({
      ok: false,
      message: 'error, contact support'
    });
  }

}


const login = async (req, res = response) => {
  try{
    const { email, password } = req.body;
    const usuario = await Usuario.findOne({email});

    if(!usuario){
      return res.status(400).json({
        ok: false,
        message: 'Usuario y contraseña no son correctos 1'
      });
    }

    // confirmar los password
    const validPassword = bcrypt.compareSync(password, usuario.password);

    if(!validPassword){
      return res.status(400).json({
        ok: false,
        message: 'Usuario y contraseña no son correctos 2'
      });
    }

    const token = await generateJWT(usuario.id, usuario.name);

    return res.status(200).json({
      ok: true,
      uid: usuario.id,
      name: usuario.name,
      message: 'login success',
      token
    });

  } catch(err){
    console.log(err);
    return res.status(500).json({
      ok: false,
      message: 'error contact support'
    });
  }
}


const refreshToken = async (req, res = response) => {
  try{

    const { uid, name } = req;
    const token = await generateJWT(uid, name);
    return res.json({ok: true, message:"refresh token", token});

  } catch( err ){
    console.log(err);
    return res.status(500).json({ok: false, message:"error"});
  }
}


module.exports = {
  register,
  login,
  refreshToken
};
