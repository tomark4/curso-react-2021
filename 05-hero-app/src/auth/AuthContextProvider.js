import { createContext, useReducer, useEffect } from 'react'
import authReducer from './authReducer'

export const AuthContext = createContext()

const init = () => {
  return JSON.parse(localStorage.getItem('user')) || {logged: false};
}

const AuthContextProvider = ({children}) => {
  const [user, dispatch] = useReducer(authReducer, {}, init)

  useEffect(() => {
    localStorage.setItem('user', JSON.stringify(user));
  },[user])

  return (
    <AuthContext.Provider value={{user, dispatch}}>
      {children}
    </AuthContext.Provider>
  )
}

export default AuthContextProvider;
