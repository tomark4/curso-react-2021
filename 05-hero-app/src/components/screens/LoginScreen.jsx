import { useContext } from "react";
import {AuthContext} from '../../auth/AuthContextProvider';
import {types} from '../../types/types';

const LoginScreen = ({history}) => {
    const {dispatch} = useContext(AuthContext);

    const lastPath = localStorage.getItem('lastPath') || "/";

    const handleLogin = () => {
        dispatch({
            type: types.login,
            payload: {
                name: 'jose'
            }
        });
        history.replace(lastPath)
    }

    return (
        <div>
            <div className="container mt-5">
                <div className="row">
                    <div className="col-6 mx-auto">
                        <div className="card">
                            <div className="card-body">
                                <h4>Login</h4>
                                <hr/>
                                <button className="btn btn-success btn-block"
                                onClick={handleLogin}>Login</button>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    )
}

export default LoginScreen
