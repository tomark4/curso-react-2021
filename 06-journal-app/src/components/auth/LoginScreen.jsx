import { useDispatch, useSelector } from 'react-redux';
import { useState } from 'react';
import { Link } from 'react-router-dom';
import { loginWithGoogle, loginWithEmailPassword } from '../../actions/auth';
import validator from 'validator';
import {
    removeError,
    setError
} from '../../actions/ui';

const LoginScreen = () => {
    const dispatch = useDispatch();
    const { ui } = useSelector( state => state );

    const [email, setEmail] = useState("test1@jotta.com");
    const [password, setPassword] = useState("123456");

    const handleSubmit = (e) => {
        e.preventDefault();
        if( !validator.isEmail(email) ){
            dispatch( setError('email invalid') );
            return false;
        }
        if( validator.isEmpty(password) || password.length < 6){
            dispatch( setError('password invalid') );
            return false;
        }
        dispatch( removeError() );
        dispatch( loginWithEmailPassword(email, password) );
    }

    const handleLoginGoogle = () => {
        dispatch( loginWithGoogle() );
    }

    return (
        <>
            <h3 className='auth__title'>Ingresar</h3>
            <form onSubmit={handleSubmit}>
                { ui?.errorMessage &&
                    <div className="auth__alert-error">{ ui?.errorMessage }</div>
                }
                <div>
                    <input
                    type="email"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={email}
                    onChange={(e) => setEmail(e.target.value)}/>
                </div>

                <div>
                    <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    value={password}
                    onChange={(e) => setPassword(e.target.value)}/>
                </div>

                <div>
                    <button type="submit" className={`btn btn-primary btn-block btn-lg ${ui.loading && 'disabled'}`}
                    disabled={ui.loading}>
                        {ui.loading ? (
                            <>
                                <i class="fa fa-spinner fa-spin"></i> Validando
                            </>
                        ) : 'Ingresar'}
                    </button>
                </div>

                <div className="google-btn" onClick={handleLoginGoogle}>
                    <div className="google-icon-wrapper">
                        <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                    </div>
                    <p className="btn-text">
                        <b>Sign in with google</b>
                    </p>
                </div>

            </form>
            <Link to="/auth/register" className="link">Â¿No tienes cuenta? Registrate</Link>
        </>
    )
}

export default LoginScreen
