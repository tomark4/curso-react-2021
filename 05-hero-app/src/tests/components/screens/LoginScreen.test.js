import LoginScreen from "../../../components/screens/LoginScreen";
import { mount } from 'enzyme';
import { AuthContext } from "../../../auth/AuthContextProvider";
import { types } from "../../../types/types";

describe('pruebas en login screen', () => {
  const context = {
    user: {
      logged: false,
    },
    dispatch: jest.fn()
  }

  const history ={
    replace: jest.fn()
  }

  const wrapper = mount(
    <AuthContext.Provider value={ context }>
      <LoginScreen history={history}/>
    </AuthContext.Provider>
  );

  test("debe de mostrarse correctamente", () => {

    expect(wrapper).toMatchSnapshot();

  });

  test('debe de realizar el dispatch y la navegacion', () => {
    wrapper.find('.btn').prop('onClick')();

    expect(context.dispatch).toHaveBeenCalledWith({
      type: types.login,
      payload: {
          name: 'jose'
      }
    })
    expect( context.dispatch ).toHaveBeenCalledTimes(1);
    expect( history.replace ).toHaveBeenCalledTimes(1)
  });


});
