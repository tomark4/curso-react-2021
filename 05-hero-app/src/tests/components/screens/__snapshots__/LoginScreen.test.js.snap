// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en login screen debe de mostrarse correctamente 1`] = `
<div>
  <div
    className="container mt-5"
  >
    <div
      className="row"
    >
      <div
        className="col-6 mx-auto"
      >
        <div
          className="card"
        >
          <div
            className="card-body"
          >
            <h4>
              Login
            </h4>
            <hr />
            <button
              className="btn btn-success btn-block"
              onClick={[Function]}
            >
              Login
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
